### PR TITLE
Support for multiple output layers

### PR DESCRIPTION
**Pull request by patrick-wilken**
This is needed to output the different factors in Factored NMT (also see #102).
I'm not familiar with all the details of search and if multiple output layers can lead to problems in some cases, but in [my case](https://github.com/patrick-wilken/returnn-examples/tree/master/factored-nmt)  it works and I'm pretty sure it does not break the one-output-layer case.

### CODE REVIEW

File: TFEngine.py [hunk 1]
```diff
@@ -1857,86 +1857,118 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
...
+                  dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx + b]),
+                  file=log.v4)
+
         if out_cache is not None:
           corpus_seq_idx = dataset.get_corpus_seq_idx(seq_idx[i])
           assert corpus_seq_idx not in out_cache
           seq_idx_to_tag[corpus_seq_idx] = seq_tag[i]
           if out_beam_size is None:
-            out_cache[corpus_seq_idx] = dataset.serialize_data(key=target_key, data=output[out_idx])
+            out_cache[corpus_seq_idx] = '\t'.join(dataset.serialize_data(key=target_keys[t],
```
Comment (patrick-wilken) [inline TFEngine.py]: I simply used tabs to separate the outputs. Maybe different files would be better?

Comment (albertz) [reply TFEngine.py]: Yes, no tabs.
But I think separate output files might make it more complicated than necessary. Why not just a dict? (But only for the case `isinstance(output_layer_names, list)`.)

File: TFEngine.py [hunk 2]
```diff
@@ -1950,7 +1991,13 @@ def extra_fetches_callback(seq_idx, seq_tag, output, targets=None, beam_scores=N
       assert len(out_cache) - 1 in out_cache
       if output_file_format == "txt":
         for i in range(len(out_cache)):
-          output_file.write("%s\n" % out_cache[i])
+          if isinstance(out_cache[i], dict):
+            # manually construct dict string to guarantee deterministic order
+            out_string = "{" + ', '.join("%r: %s" % (output_layer_names[t], out_cache[i][output_layer_names[t]]) \
+                for t in range(num_serializable_targets)) + "}\n"
```
Comment (albertz) [inline TFEngine.py]: There is `Util.betterRepr` for this purpose.
Also then you don't need the comment (as it becomes clear when you use `Util.betterRepr`).

Comment (albertz) [reply TFEngine.py]: However, I think it makes more sense to just don't allow txt format output for this case anyway. That makes it just more complicated to parse it later. As said: Don't make it more complicated than necessary.

Comment (albertz) [reply TFEngine.py]: Note that you should use `Util.betterRepr` for the py format case, though.

File: TFEngine.py [hunk 3]
```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
...
+            if out_cache is not None:
+              if out_beam_sizes[t] is None:
+                  out_data = dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx])
+              else:
+                assert beam_scores[t] is not None
+                out_data = [
+                    (beam_scores[t][i][b], dataset.serialize_data(key=target_keys[t], data=outputs[t][out_idx + b]))
+                    for b in range(out_beam_sizes[t])]
+
+              if num_serializable_targets == 1:
```
Comment (albertz) [inline TFEngine.py]: The check should be `isinstance(output_layer_names, dict)`, and not `num_serializable_targets == 1`.

File: TFEngine.py [hunk 4]
```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
...
-                "  hyp %i, score %f:" % (b, beam_scores[i][b]),
-                dataset.serialize_data(key=target_key, data=output[out_idx + b]),
-                file=log.v4)
         if out_cache is not None:
           corpus_seq_idx = dataset.get_corpus_seq_idx(seq_idx[i])
           assert corpus_seq_idx not in out_cache
           seq_idx_to_tag[corpus_seq_idx] = seq_tag[i]
-          if out_beam_size is None:
-            out_cache[corpus_seq_idx] = dataset.serialize_data(key=target_key, data=output[out_idx])
+          if num_serializable_targets > 1:
```
Comment (albertz) [inline TFEngine.py]: The check should be `isinstance(output_layer_names, dict)`, and not `num_serializable_targets == 1`.

File: TFEngine.py [hunk 5]
```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
...
+        output_layer_beam_scores.append(output_layer.get_search_choices().beam_scores)
+      out_beam_sizes.append(output_layer.output.beam_size)
+      target_keys.append(output_layer.target or self.network.extern_data.default_target)
 
     out_cache = None
     seq_idx_to_tag = {}
     if output_file:
       assert output_file_format in {"txt", "py"}
-      assert dataset.can_serialize_data(target_key)
+      num_serializable_targets = sum(dataset.can_serialize_data(target_key) for target_key in target_keys)
```
Comment (albertz) [inline TFEngine.py]: Why so complicated? Just assert that all can be serialized.

File: TFEngine.py [hunk 6]
```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
...
-      print("Given output %r is after decision (no beam)." % output_layer, file=log.v1)
-    else:
-      print("Given output %r has beam size %i." % (output_layer, out_beam_size), file=log.v1)
-      output_layer_beam_scores = output_layer.get_search_choices().beam_scores
-    target_key = output_layer.target or self.network.extern_data.default_target
+    if not isinstance(output_layer_names, list):
+      output_layer_names = [output_layer_names]
+    num_targets = len(output_layer_names)
+
+    output_layers, out_beam_sizes, output_layer_beam_scores, target_keys = [], [], [], []
```
Comment (albertz) [inline TFEngine.py]: There should be a comment about the size of the lists.

File: TFEngine.py [hunk 7]
```diff
@@ -1857,86 +1857,127 @@ def search(self, dataset, do_eval=True, output_layer_name="output", output_file=
...
     if output_file:
       assert output_file_format in {"txt", "py"}
-      assert dataset.can_serialize_data(target_key)
+      num_serializable_targets = sum(dataset.can_serialize_data(target_key) for target_key in target_keys)
+      assert num_serializable_targets > 0
       assert not os.path.exists(output_file)
       print("Will write outputs to: %s" % output_file, file=log.v2)
       output_file = open(output_file, "w")
-      out_cache = {}  # corpus-seq-idx -> str|list[(float,str)]
+      out_cache = {}  # corpus-seq-idx -> str|list[(float,str)]|dict[str -> str|list[(float,str)]] depending on number of output layers and beam sizes
```
Comment (albertz) [inline TFEngine.py]: It should not depend on number of output layers.

Comment (albertz) [reply TFEngine.py]: Also not on beam sizes. It actually depends if the output is after a choice (i.e. single best) or before (the whole beam).

File: TFEngine.py (hunk 5 above)
Comment (patrick-wilken) [reply TFEngine.py]: Before, it was possible to do search with a non-serializable target when no output file was given. That just prints the output indices to the console (if verbosity >=4). I wanted to keep that behaviour even if other targets are written to the output file. But I agree, it is not super useful, so I can remove it.

Comment (albertz) [reply TFEngine.py]: But where is the problem? That would still be the case (this is inside the `if output_file:` branch). Just do `assert all([dataset.can_serialize_data(target_key) for target_key in target_keys])`.

Comment (patrick-wilken) [reply TFEngine.py]: Yes, I will. I just mean, then it is not possible to have a console output for a non-serializable feature A while writing feature B to file.

File: TFEngine.py (hunk 3 above)
Comment (patrick-wilken) [reply TFEngine.py]: list you mean? But I convert output_layer_names to list if it is a string, should I change that?

Comment (albertz) [reply TFEngine.py]: Sorry, yes, I meant list. You can also do `want_output_list = isinstance(output_layer_names, list)` in the beginning. Or do not convert `output_layer_names`. Whatever makes the code simpler.

File: TFEngine.py (hunk 6 above)
Comment (patrick-wilken) [reply TFEngine.py]: Oh, forgot that one. Do you mean just a comment saying all lists will be of length num_targets? Or do you mean the data type?

Comment (albertz) [reply TFEngine.py]: The data types would probably be clearer (it's not totally obvious in all the cases), but also the length. So maybe make each assign an individual line also.

**Comment by patrick-wilken** [comment]
Ok, I think I have addressed all the requests now.

**Pull request merged**